### PR TITLE
Multi-index indexing

### PR DESCRIPTION
**Pull request by benbovy**
Follows #767.

This is incomplete (it still needs some tests and documentation updates), but it is working for both `Dataset` and `DataArray` objects. I also don't know if it is fully compatible with lazy indexing (Dask). 

Using the example from #767:

```
In [4]: da.sel(band_wavenumber={'band': 'foo'})
Out[4]:
<xarray.DataArray (wavenumber: 2)>
array([ 0.00017,  0.00014])
Coordinates:
  * wavenumber  (wavenumber) float64 4.05e+03 4.05e+03
```

As shown in this example, similarily to pandas, it automatically renames the dimension and assigns a new coordinate when the selection doesn't return a `pd.MultiIndex` (here it returns a `pd.FloatIndex`).

In some cases this behavior may be unwanted (??), so I added a `drop_level` keyword argument (if `False` it keeps the multi-index and doesn't change the dimension/coordinate names):

```
In [5]: da.sel(band_wavenumber={'band': 'foo'}, drop_level=False)
Out[5]:
<xarray.DataArray (band_wavenumber: 2)>
array([ 0.00017,  0.00014])
Coordinates:
  * band_wavenumber  (band_wavenumber) object ('foo', 4050.2) ('foo', 4050.3)
```

Note that it also works with `DataArray.loc`, but (for now) in that case it always returns the multi-index:

```
In [6]: da.loc[{'band_wavenumber': {'band': 'foo'}}]
Out[6]:
<xarray.DataArray (band_wavenumber: 2)>
array([ 0.00017,  0.00014])
Coordinates:
  * band_wavenumber  (band_wavenumber) object ('foo', 4050.2) ('foo', 4050.3)
```

This is however inconsistent with `Dataset.sel` and `Dataset.loc` that both apply `drop_level=True` by default, due to their different implementation. Two solutions: (1) make `DataArray.loc` apply drop_level by default, or (2) use `drop_level=False` by default everywhere.

### CODE REVIEW

**Comment by shoyer** [comment]
This looks very nice! I would probably opt for making `drop_level=True` the default for `.loc` -- it's hard to see when that wouldn't be desirable. In fact, I'm not even sure that it's worth having the option to set `drop_level=False`.

**Comment by benbovy** [comment]
I followed your suggestions.

Two more comments (not critical issues I think) :

- Multi-index level drop only works when a dict is provided for a dimension, i.e., `da.sel(band_wavenumber='foo')` still returns the full `pandas.MultiIndex`.  This would require for `indexing.convert_label_indexer` to return an updated index also when `label` is not dict-like, which is not that straightforward I think.
- There is a potential conflict when providing a dictionary to `DataArray.loc`, it may be interpreted either as a mapping of dimensions / labels or a mapping of index levels / labels for the 1st dimension. For now, the second option is not handled, e.g., `da.loc[{'band': 'foo'}]` raises a `KeyError`.

In summary, `da.loc[{'band_wavenumber': {'band': 'foo'}}]` returns a DataArray with updated index and coordinate / dimension name, `da.loc['foo']` returns a DataArray with the full multi-index, and `da.loc[{'band': 'foo'}]` raises a `KeyError`. Three different results for the same operation, although I think that the first one is more explicit and better.

**Comment by shoyer** [comment]
> Multi-index level drop only works when a dict is provided for a dimension, i.e., da.sel(band_wavenumber='foo') still returns the full pandas.MultiIndex. This would require for indexing.convert_label_indexer to return an updated index also when label is not dict-like, which is not that straightforward I think.

Indeed. This would require an another data structure somewhere keeping track of level names -- and ideally also ensuring that they are always unique (like dimensions). This seems fine to me for now.

> There is a potential conflict when providing a dictionary to DataArray.loc, it may be interpreted either as a mapping of dimensions / labels or a mapping of index levels / labels for the 1st dimension. For now, the second option is not handled, e.g., da.loc[{'band': 'foo'}] raises a KeyError.

I agree -- better to require the user to be explicit. I also don't see many use cases for specifying the coordinate value and level name but not the dimension name. What happens if you type `da.loc[{'band': 'foo'}, :]` or `da.loc[{'band': 'foo'}, ...]`? In principle these are not ambiguous, but I could only guess at what happens without looking at the code.

**Comment by benbovy** [comment]
After refactoring a bit, `da.loc[{'band': 'foo'}, :]` and `da.loc[{'band': 'foo'}, ...]` now work as expected. 

`da.loc[{'band': 'foo'}]` still raises a `KeyError`. Although it is possible to catch this exception, it still remains ambiguous so it is probably better to do nothing.

**Comment by shoyer** [comment]
Dictionaries not hashable, so we might be able to detect this case by
calling hash() and raising a KeyError with alternative suggestions. Just an
idea -- not sure if there's a clean way to implement that.

**Comment by benbovy** [comment]
Unless you see any other issues, I think that this feature doesn't need more development for now.

I'll be back next week to finish this PR (write some tests and doc).

**Comment by tippetts** [comment]
This will be a great feature. I for one am really looking forward to using it.

Will this work also allow saving to/reading from hdf5 and netcdf files with a MultiIndex? If not, can you give a sketch outline of the approach you (Stephan or Benoit) would take? I assume it would involve saving the information about the MultiIndex structure in some transformed way that fits into an hdf5 file, then reconstructing it on the read. I might need to hack together something for that before MultiIndex serialization makes it into xarray, but I'd like to make sure I don't veer too far off from the real solution that will ultimately come out.

**Comment by benbovy** [comment]
I hope to have some time next week to work again on this PR.

@tippetts  You can see in #719  a few comments about saving/reading xarray data objects with multi-index to/from netCDF. I also looking forward to see this feature implemented - actually I need it for another project that uses xarray - so maybe I'll find some time in the next couple of weeks to start a new PR on this.

**Comment by benbovy** [comment]
I finally managed to add some tests and docs.

Two more comments:

- Index replacement: if the multi-index has unnamed levels, the chosen name for the replaced index (if any) is `<old_dim>_unnamed_level` (`old_dim` is the name of the replaced multi-index dimension).
Not sure whether or not it is really necessary to track the index level number. Single indexes returned by `get_loc_level` don't keep that information. Anyway, it is always better to set explicit names for the index levels.
- Indexing on multi-index now also works with nested tuples (#850).

@shoyer I think that it is ready for review. I successfully run the tests on my local branch. Currently, CI tests seem broken for some reason I don't know.

File: xarray/core/dataarray.py [hunk 1]
```diff
@@ -244,6 +240,16 @@ def _replace_maybe_drop_dims(self, variable, name=__default):
                                  if set(v.dims) <= allowed_dims)
         return self._replace(variable, coords, name)
 
+    def _replace_indexes(self, indexes):
+        obj = self
+        for dim, idx in iteritems(indexes):
+            if idx.name is None:
+                idx.name = dim + "_unnamed_level"
```
Comment (shoyer) [inline xarray/core/dataarray.py]: Maybe should handle this instead by requiring names for all levels on multi-indexes used by xarray? Or we could pick defaults (e.g., `level_0`, `level_1`, etc.). I worry that this choice is likely to give you multiple levels with the same name "unnamed_level", though I suppose this isn't so bad given that the unnamed level stuff is mostly there as a soft reminder to add names.

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -244,6 +240,16 @@ def _replace_maybe_drop_dims(self, variable, name=__default):
...
+            obj = obj.rename({dim: idx.name})
+            new_coord = Coordinate(idx.name, idx)
+            obj = obj._replace(coords={idx.name: new_coord})
```
Comment (shoyer) [inline xarray/core/dataarray.py]: It's better to accumulate all the new coordinates in a dictionary, and then call `_replace` once outside the loop.

File: doc/indexing.rst [hunk 3]
```diff
@@ -294,6 +294,37 @@ elements that are fully masked:
...
+    da_midx
+    da_midx.sel(x={'one': 'a', 'two': 0})
+    da_midx.loc[{'one': 'a'}, ...]
+
+As shown in the last example above, xarray handles partial selection on
+pandas multi-index ; it automatically renames the dimension and replaces the
+coordinate when a single index is returned (level drop).
+
+Like pandas, it is also possible to use tuples of tuples, lists or slices
+(in that case xarray always returns the full multi-index):
```
Comment (shoyer) [inline doc/indexing.rst]: Is this the preferred behavior or just what you could get to work? i.e., will we want to change it in the future?

Comment (benbovy) [inline doc/indexing.rst]: Just what I could get to work...

File: xarray/core/dataarray.py (hunk 1 above)
Comment (benbovy) [inline xarray/core/dataarray.py]: IMO `level_0`, `level_1`, etc. is the best option. I finally found a way to handle this: set default levels names on a copy of the MultIndex before calling `convert_label_indexer`.

**Comment by benbovy** [comment]
Thanks for the review @shoyer !

I made some updates according to your comments.

A possible remaining issue is that using a tuple of tuples, lists or slices exactly behaves like pandas (it calls `get_locs`) while a nested list can still be used to select couples of multi-index labels (it calls `get_indexer`). Should we allow only one of two behaviors, or keep both (and mention it in the docs)?

**Comment by shoyer** [comment]
> A possible remaining issue is that using a tuple of tuples, lists or slices exactly behaves like pandas (it calls get_locs) while a nested list can still be used to select couples of multi-index labels (it calls get_indexer). Should we allow only one of two behaviors, or keep both (and mention it in the docs)?

Can you give a quick example of what using the nested list look like?

**Comment by benbovy** [comment]
```
>>> index = pd.MultiIndex.from_product((list('abc'), range(3)))
>>> index.names = ('one', 'two')
>>> data = xr.DataArray(range(9), [('x', index)])
>>> data
<xarray.DataArray (x: 9)>
array([0, 1, 2, 3, 4, 5, 6, 7, 8])
Coordinates:
  * x        (x) object ('a', 0) ('a', 1) ('a', 2) ('b', 0) ('b', 1) ...
```

Example using a tuple of scalar/tuple/list/slice labels for each level:

```
>>> data.sel(x=('a', (0, 1)))
<xarray.DataArray (x: 2)>
array([0, 1])
Coordinates:
  * x        (x) object ('a', 0) ('a', 1)
```

Example using a list of several combinations of multi-index labels:  

```
>>> data.sel(x=[('a', 0), ('b', 1), ('c', 1)])
<xarray.DataArray (x: 3)>
array([0, 4, 7])
Coordinates:
  * x        (x) object ('a', 0) ('b', 1) ('c', 1)
```

All the following examples raise an error:

```
>>> data.sel(x=(('a', 0), ('b', 1), ('c', 1)))
>>> data.sel(x=['a', (0, 1)])
>>> data.sel(x=[['a', 0], ['b', 1], ['c', 1]])
```

Note that after verification, pandas has the same behavior for all of the examples shown here.

**Comment by shoyer** [comment]
Yeah, that behavior looks pretty good to me -- there are limits on how flexible you can make things :).

We are fortunate that we have `.sel`, which lets us side-step some tricky ambiguous cases. Still, how do we handle `data.loc['a', 0]` if `data` is 2-dimensional? Pandas uses some terrible fall-back logic to allow the `0` to index either the second level or the second dimension, depending upon which works.

**Comment by benbovy** [comment]
That behavior looks good enough to me too! It is indeed already quite flexible :-).

Do you think that we really need to handle `data.loc['a', 0]` if `data` is 2 or higher dimensional? We would also have to handle the case `data.loc[{'one': 'a', 'two': 0}]`...
Having to explicitly write `data.loc[('a', 0), :]` or `data.loc[('a', 0), ...]` seems fine to me (at least for now).

**Comment by shoyer** [comment]
I don't think we need to handle it -- I'm definitely supportive of only supporting unambiguous syntax. Let's just note the difference from pandas.

**Comment by benbovy** [comment]
Noted the difference in documentation.

Hope that the whole "multi-level indexing" section is clear enough, not sure if it's proper English.

File: doc/indexing.rst [hunk 4]
```diff
@@ -294,6 +294,60 @@ elements that are fully masked:
...
+    idx = pd.MultiIndex.from_product([list('abc'), [0, 1]],
+                                     names=('one', 'two'))
+    da_midx = xr.DataArray(np.random.rand(6, 3),
+                           [('x', idx), ('y', range(3))])
+    da_midx
+    da_midx.sel(x={'one': 'a', 'two': 0})
+    da_midx.loc[{'one': 'a'}, ...]
+
+As shown in the last example above, xarray handles partial selection on
+pandas multi-index ; it automatically renames the dimension and replaces the
```
Comment (shoyer) [inline doc/indexing.rst]: nit: I would use a new sentence instead of the semicolon.

File: doc/indexing.rst [hunk 5]
```diff
@@ -294,6 +294,60 @@ elements that are fully masked:
...
+   da_midx.sel(x=[('a', 0), ('b', 1)])
+
+A single, flat tuple can be used to select a given combination of
+multi-index labels:
+
+.. ipython:: python
+
+   da_midx.sel(x=('a', 0))
+
+Unlike pandas, xarray can't make the distinction between index levels and
```
Comment (shoyer) [inline doc/indexing.rst]: Instead of "can't make the distinction", let's say "does not guess".

File: doc/whats-new.rst [hunk 6]
```diff
@@ -39,10 +39,14 @@ Enhancements
   attributes are retained in the resampled object. By
   `Jeremy McGibbon <https://github.com/mcgibbon>`_.
 
+- DataArray and Dataset methods :py:meth:`sel` and :py:meth:`loc` now
+  accept dictionaries or nested tuples for indexing on multi-index dimensions.
+  By `Benoit Bovy <https://github.com/benbovy>`_.
```
Comment (shoyer) [inline doc/whats-new.rst]: Can you also please add a note about the changed behavior (we now drop levels, which is consistent with pandas) in the "Breaking changes" section above?

Also, add a reference here to the documentation section you added:

    :ref:`multi-level indexing`

File: xarray/core/dataarray.py [hunk 7]
```diff
@@ -86,24 +86,22 @@ def __init__(self, data_array):
...
-            return tuple(lookup_positions(dim, labels) for dim, labels
-                         in zip(self.data_array.dims, key))
+            labels = indexing.expanded_indexer(key, self.data_array.ndim)
+            key = dict(zip(self.data_array.dims, labels))
+        return indexing.remap_label_indexers(self.data_array, key)
 
     def __getitem__(self, key):
-        return self.data_array[self._remap_key(key)]
+        pos_indexers, new_indexes = self._remap_key(key)
+        ds = self.data_array[pos_indexers]._to_temp_dataset()
```
Comment (shoyer) [inline xarray/core/dataarray.py]: Can we avoid creating the temporary dataset here? `_to_temp_dataset()`/`_from_temp_dataset` is a little expensive (and also a bit of a hack).

File: xarray/core/dataset.py [hunk 8]
```diff
@@ -419,6 +419,18 @@ def _replace_vars_and_dims(self, variables, coord_names=None,
             obj = self._construct_direct(variables, coord_names, dims, attrs)
         return obj
 
+    def _replace_indexes(self, indexes):
+        variables = OrderedDict()
+        for k, v in iteritems(self._variables):
+            if k in indexes.keys():
+                idx = indexes[k]
+                variables[k] = Coordinate(idx.name, idx)
```
Comment (shoyer) [inline xarray/core/dataset.py]: If `idx.name != k` above, then this could be constructing an invalid dataset.

I think we should create `Coordinate(k, idx)` and then remap back to the original names below, if necessary.

File: xarray/core/dataset.py [hunk 9]
```diff
@@ -419,6 +419,18 @@ def _replace_vars_and_dims(self, variables, coord_names=None,
...
+            else:
+                variables[k] = v
+        obj = self._replace_vars_and_dims(variables)
+        dim_names = {dim: idx.name for dim, idx in iteritems(indexes)}
+        return obj.rename(dim_names)
```
Comment (shoyer) [inline xarray/core/dataset.py]: Can we make the rename only done if necessary? I think this can be kind of expensive.

Putting things together:
```python
variables = self._variables.copy()
for name, idx in indexes.items():
   variables[name] = Coordinate(name, idx)
obj = self._replace_vars_and_dims(variables)

# switch from dimension to level names, if necessary
dim_names = {}
for dim, idx in indexes.items():
    if idx.name != dim:
        dim_names[dim] = idx.name
if dim_names:
    obj = obj.rename(dim_name)
```

File: xarray/core/dataarray.py (hunk 7 above)
Comment (benbovy) [inline xarray/core/dataarray.py]: Yeah I agree this is not very nice. I did this to avoid duplicating the `_replace_indexes` method, but I agree that it may be actually the least bad option.

File: xarray/core/indexing.py [hunk 10]
```diff
@@ -177,18 +206,36 @@ def convert_label_indexer(index, label, index_name='', method=None,
...
-                                            dim, method, tolerance))
-                for dim, label in iteritems(indexers))
+
+    pos_indexers, new_indexes = {}, {}
+    for dim, label in iteritems(indexers):
+        index = data_obj[dim].to_index()
+
+        if isinstance(index, pd.MultiIndex):
+            # set default names for multi-index unnamed levels so that
+            # we can safely rename dimension / coordinate later
```
Comment (shoyer) [inline xarray/core/indexing.py]: This looks great! We might also consider moving this logic to [around this line](https://github.com/pydata/xarray/blob/2448b5b3fa3647419df8f14d60f1c32b9d002af7/xarray/core/variable.py#L107) of `core.variable.as_compatible_data`. All data passed to xarray objects goes through this method.

File: xarray/core/dataset.py [hunk 11]
```diff
@@ -1114,8 +1130,9 @@ def sel_points(self, dim='points', method=None, tolerance=None,
         Dataset.isel_points
         DataArray.sel_points
         """
-        pos_indexers = indexing.remap_label_indexers(
-            self, indexers, method=method, tolerance=tolerance)
+        pos_indexers, new_indexes = indexing.remap_label_indexers(
```
Comment (shoyer) [inline xarray/core/dataset.py]: if we ignore `new_indexes` here (which is fine, certainly for now), use the variable name `_` instead to indicate that it's unused.

File: xarray/core/indexing.py [hunk 12]
```diff
@@ -135,11 +135,27 @@ def _asarray_tuplesafe(values):
     return result
 
 
+def _is_nested_tuple(tup, index):
+    """Check for a compatible nested tuple and multiindex (taken from
+    pandas.core.indexing.is_nested_tuple).
```
Comment (shoyer) [inline xarray/core/indexing.py]: I'm still trying to wrap my head around exactly what this check does :).

File: xarray/core/dataset.py [hunk 13]
```diff
@@ -972,8 +986,10 @@ def sel(self, method=None, tolerance=None, **indexers):
         Dataset.isel_points
         DataArray.sel
         """
-        return self.isel(**indexing.remap_label_indexers(
-            self, indexers, method=method, tolerance=tolerance))
+        pos_indexers, new_indexes = indexing.remap_label_indexers(
+            self, indexers, method=method, tolerance=tolerance
+        )
+        return self.isel(**pos_indexers)._replace_indexes(new_indexes)
```
Comment (shoyer) [inline xarray/core/dataset.py]: Does this handle the case where `new_indexes` is None?

Comment (shoyer) [inline xarray/core/dataset.py]: Nevermind, that can't happen.

File: xarray/core/dataset.py (hunk 9 above)
Comment (benbovy) [inline xarray/core/dataset.py]: Seems much nicer!

What about

```
if not len(indexes):
    return self
```

at the top of the function, given that in many use cases indexes will be empty?  (I don't know if `self._replace_vars_and_dims` is expensive)

File: xarray/core/indexing.py [hunk 14]
```diff
@@ -177,18 +206,36 @@ def convert_label_indexer(index, label, index_name='', method=None,
...
+            # we can safely rename dimension / coordinate later
+            valid_level_names = [name or '{}_level_{}'.format(dim, i)
+                                 for i, name in enumerate(index.names)]
+            index = index.copy()
+            index.names = valid_level_names
+
+        idxr, new_idx = convert_label_indexer(index, label,
+                                              dim, method, tolerance)
+        pos_indexers[dim] = idxr
+        if new_idx is not None and not isinstance(new_idx, pd.MultiIndex):
```
Comment (shoyer) [inline xarray/core/indexing.py]: What should happen if `new_idx` *is* a MultiIndex? Right not it looks like it gets thrown away?

Comment (shoyer) [inline xarray/core/indexing.py]: We definitely need to add a test for this situation (e.g., 3 level index -> 2 level index).

File: xarray/core/indexing.py (hunk 12 above)
Comment (benbovy) [inline xarray/core/indexing.py]: So I'm ! I've just stolen this from pandas without much modification :).

Comment (shoyer) [inline xarray/core/indexing.py]: This is such a weird function the way it's currently written. Why not make this:
```python
def _is_nested_tuple(possible_tuple):
    return (isinstance(possible_tuple, tuple)
            and any(isinstance(value, (tuple, list, slice)
                    for value in possible_tuple))
```

The `isinstance(index, pd.MultiIndex)` should be outside the function, since it's totally unrelated.

File: xarray/core/indexing.py [hunk 15]
```diff
@@ -166,6 +184,17 @@ def convert_label_indexer(index, label, index_name='', method=None,
...
+    elif is_dict_like(label):
+        if not isinstance(index, pd.MultiIndex):
+            raise ValueError('cannot use a dict-like object for selection on a '
+                             'dimension that does not have a MultiIndex')
+        indexer, new_index = index.get_loc_level(tuple(label.values()),
+                                                 level=tuple(label.keys()))
+
+    elif _is_nested_tuple(label, index):
+        indexer = index.get_locs(label)
+
```
Comment (shoyer) [inline xarray/core/indexing.py]: I think we could reproduce what pandas does in terms of collapsing tuple levels if we call `get_loc_level` for non-nested tuples:

```python
# untested!

elif isinstance(label, tuple) and isinstance(index, pd.MultiIndex):
    if _is_nested_tuple(label):
        indexer = index.get_locs(label)
    else:
        indexer, new_index = index.get_loc_level(label, level=range(len(label)))

```

**Comment by shoyer** [comment]
Just went through and gave another full review -- this is looking quite nice, just a few more things to clean up!

File: xarray/core/dataset.py (hunk 9 above)
Comment (shoyer) [inline xarray/core/dataset.py]: Yes, that's even better!

`self._replace_vars_and_dims` is pretty cheap (it skips all validation) but not free.

File: xarray/core/indexing.py (hunk 15 above)
Comment (benbovy) [inline xarray/core/indexing.py]: Yes it works!

However, using non-nested tuples here consists of selecting single elements and raises the question of how we handle returned scalar values. In that specific case we should drop the dimension but keep the 0-d (multi-level) coordinate so that `data.sel(x=('a', 0))` is equivalent to `data.isel(x=0)` in the example above.

More generally, I think we definitely need to carefully address level drop in all cases.

Comment (shoyer) [inline xarray/core/indexing.py]: Doing some tests, it seems like `get_loc_level` properly drops levels from a MultiIndex, but doesn't handle dropping the level entirely. For that, we need to use `get_loc`.

Good:
```
In [11]: idx = pd.MultiIndex.from_product([['a', 'b'], [1, 2], [-1, -2]])

In [12]: idx.get_loc_level(('a', 1), [0, 1])
Out[12]:
(array([ True,  True, False, False, False, False, False, False], dtype=bool),
 Int64Index([-1, -2], dtype='int64'))

In [13]: idx.get_loc_level(('a', -1), [0, 2])
Out[13]:
(array([ True, False,  True, False, False, False, False, False], dtype=bool),
 Int64Index([1, 2], dtype='int64'))

In [15]: idx.get_loc_level(('a',), [0])
Out[15]:
(array([ True,  True,  True,  True, False, False, False, False], dtype=bool),
 MultiIndex(levels=[[1, 2], [-2, -1]],
            labels=[[0, 0, 1, 1], [1, 0, 1, 0]]))

In [16]: idx.get_loc_level((1,), [1])
Out[16]:
(array([ True,  True, False, False,  True,  True, False, False], dtype=bool),
 MultiIndex(levels=[['a', 'b'], [-2, -1]],
            labels=[[0, 0, 1, 1], [1, 0, 1, 0]]))
```

Bad:
```
In [14]: idx.get_loc_level(('a', 1, -1), [0, 1, 2])
Out[14]:
(array([ True, False, False, False, False, False, False, False], dtype=bool),
 MultiIndex(levels=[['a', 'b'], [1, 2], [-2, -1]],
            labels=[[0], [0], [1]]))
```

Good:
```
In [17]: idx.get_loc(('a', 1, -1))
Out[17]: 0
```

Comment (shoyer) [inline xarray/core/indexing.py]: So I guess we need to check the length of the tuple (probably also in the `dict` path above):
```python
elif isinstance(label, tuple) and isinstance(index, pd.MultiIndex):
    if _is_nested_tuple(label):
        indexer = index.get_locs(label)
    elif len(label) == index.nlevels:
        indexer = index.get_loc(label)
    else:
        indexer, new_index = index.get_loc_level(label, level=range(len(label)))
```

Comment (benbovy) [inline xarray/core/indexing.py]: I thinking about something like this:

```python
def _maybe_drop_levels(index):
    drop_levels = [i for i, lab in enumerate(index.labels)
                   if not np.ptp(lab.values())]
    if len(drop_levels) < len(index.labels):
        return index.droplevel()
    else:
        return index

def convert_label_indexer(...):
    
    # ...

    if isinstance(new_index, pd.MultiIndex):
        new_index = _maybe_drop_levels(new_index)

    return indexer, new_index
```

File: xarray/core/indexing.py (hunk 14 above)
Comment (benbovy) [inline xarray/core/indexing.py]: Yes multi-indexes are not updated, but maybe we should do so (see my comment below on level drop).

File: xarray/core/indexing.py (hunk 15 above)
Comment (shoyer) [inline xarray/core/indexing.py]: The advantage of doing something like my proposed logic (which I think is similar to what pandas does) is that whether a level is dropped depends only on the indexer type and the number of multi-index levels, as opposed to dropping levels in a way that depends also on the particular values in the indexer and index. Code that depends only on type information rather than values is generally easier to understand and less error prone.

Comment (shoyer) [inline xarray/core/indexing.py]: also, `np.ptp` only works on numbers, not strings.

Comment (benbovy) [inline xarray/core/indexing.py]: Yep, I used `np.ptp` here as pandas `MultiIndex.labels` elements are integers by definition.

Anyway, I get your logic. It is also much more efficient!

File: xarray/core/indexing.py (hunk 10 above)
Comment (benbovy) [inline xarray/core/indexing.py]: Shouldn't be better to move this logic to `indexing.PandasIndexAdapter`, wrap the `names` (and/or `name`) property, and return a `PandasIndexAdapter` object from `convert_label_indexer`?

This is because I worry about implicit copy or renaming.

The problem would be to set default level names that are unique across dimensions, but maybe we can pass the variable name in the `PandasIndexAdapter` constructor (as an additional `name` kwarg).

Comment (benbovy) [inline xarray/core/indexing.py]: Or maybe moving this [here](https://github.com/pydata/xarray/blob/2448b5b3fa3647419df8f14d60f1c32b9d002af7/xarray/core/variable.py#L1093) (`to_index()`) would be simpler...

Comment (shoyer) [inline xarray/core/indexing.py]: `PandasIndexAdapter` is really for handling aspects of pandas indexes that we can't change by converting into a new index. For example, PeriodIndex claims to have `dtype=int`, when it's really a sort of object dtype.

`pandas.Index` is actually immutable, so we don't need to worry about changing the data. Calling `.copy()` just copies the metadata and reuses the same data. That said, calling `.set_names()` (which returns a new MultiIndex) is probably a more obvious way of handling this.

It does looks like we already some similar logic in `to_index()`, so that could be a reasonable place to put this, too.

Comment (benbovy) [inline xarray/core/indexing.py]: Compared to `as_compatible_data`, moving to `to_index()` would create a copy of the multi-index (almost free I guess) each time it is called but it would also allow tracking level name changes on the original `MultiIndex`:

```python
idx = pd.MultiIndex.from_product([['a', 'b'], [1, 2], [-1, -2]])
>>> y = xr.DataArray(np.random.rand(2 * 2 * 2), [('x', idx)])
>>> y.x.to_index()
MultiIndex(levels=[['a', 'b'], [1, 2], [-2, -1]],
           labels=[[0, 0, 0, 0, 1, 1, 1, 1], [0, 0, 1, 1, 0, 0, 1, 1], [1, 0, 1, 0, 1, 0, 1, 0]],
           names=['x_level_0', 'x_level_1', 'x_level_2'])
>>> idx.names = ('one', 'two', 'three')
>>> y.x.to_index()
MultiIndex(levels=[['a', 'b'], [1, 2], [-2, -1]],
           labels=[[0, 0, 0, 0, 1, 1, 1, 1], [0, 0, 1, 1, 0, 0, 1, 1], [1, 0, 1, 0, 1, 0, 1, 0]],
           names=['one', 'two', 'three'])
```

and we have also direct access to the dimension/coordinate name so that we can set default level names that are unique (as shown above).

Comment (shoyer) [inline xarray/core/indexing.py]: > and we have also direct access to the dimension/coordinate name so that we can set default level names that are unique (as shown above).

True, but unless we allow directly accessing levels as variables, `level_0` on `dim_0` is an unambiguous as `dim_0_level_0`.

**Comment by benbovy** [comment]
Thanks for your second review @shoyer ! It actually helped me to better understand the indexing logic of pandas (never too late)!

I made some updates according to your comments.

I think we're getting closer to a working feature!

File: doc/data-structures.rst [hunk 16]
```diff
@@ -115,11 +115,9 @@ If you create a ``DataArray`` by supplying a pandas
     df
     xr.DataArray(df)
 
-xarray does not (yet!) support labeling coordinate values with a
-:py:class:`pandas.MultiIndex` (see :issue:`164`).
-However, the alternate ``from_series`` constructor will automatically unpack
-any hierarchical indexes it encounters by expanding the series into a
-multi-dimensional array, as described in :doc:`pandas`.
+Xarray supports labeling coordinate values with a :py:class:`pandas.MultiIndex`.
```
Comment (shoyer) [inline doc/data-structures.rst]: It might make sense simply to drop this paragraph instead -- do we really need to explicitly call out MultiIndex if it's supported?

File: xarray/core/dataarray.py [hunk 17]
```diff
@@ -244,6 +239,23 @@ def _replace_maybe_drop_dims(self, variable, name=__default):
                                  if set(v.dims) <= allowed_dims)
         return self._replace(variable, coords, name)
 
+    def _replace_indexes(self, indexes):
+        if not len(indexes):
```
Comment (shoyer) [inline xarray/core/dataarray.py]: just use `if not indexes:` (it's more idiomatic)

File: xarray/core/variable.py [hunk 18]
```diff
@@ -1096,7 +1096,13 @@ def to_index(self):
         # basically free as pandas.Index objects are immutable
         assert self.ndim == 1
         index = self._data_cached().array
-        if not isinstance(index, pd.MultiIndex):
+        if isinstance(index, pd.MultiIndex):
+            # set default names for multi-index unnamed levels so that
+            # we can safely rename dimension / coordinate later
+            valid_level_names = [name or '{}_level_{}'.format(self.name, i)
```
Comment (shoyer) [inline xarray/core/variable.py]: I would lean towards defaulting to simply "level_0",  etc., but I don't feel too strongly here.

**Comment by shoyer** [comment]
I did a little bit of testing. Here is one case I found where things don't work as I expected:
```
In [32]: idx = pd.MultiIndex.from_product([['a', 'b'], [0, 1, 2]], names=['lev0', 'lev1'])

In [33]: data = xr.DataArray(range(6), [('x', idx)])

In [34]: data.sel(x=('a',))
Out[34]:
<xarray.DataArray (lev1: 3)>
array([0, 1, 2])
Coordinates:
  * lev1     (lev1) int64 0 1 2

In [35]: data.sel(x='a')
Out[35]:
<xarray.DataArray (x: 3)>
array([0, 1, 2])
Coordinates:
  * x        (x) object ('a', 0) ('a', 1) ('a', 2)
```

I would expect an index drop in the last case, too. I guess we need to check for scalars.

File: xarray/core/indexing.py [hunk 19]
```diff
@@ -166,6 +175,28 @@ def convert_label_indexer(index, label, index_name='', method=None,
...
+        elif len(label) == index.nlevels:
+            indexer = index.get_loc(label)
+        else:
+            indexer, new_index = index.get_loc_level(
+                label, level=list(range(len(label)))
+            )
+
     else:
         label = _asarray_tuplesafe(label)
         if label.ndim == 0:
```
Comment (shoyer) [inline xarray/core/indexing.py]: this is where scalars end up -- probably need to add a clause here to handle MultiIndex

File: doc/data-structures.rst (hunk 16 above)
Comment (benbovy) [inline doc/data-structures.rst]: No I don't think we need it. However, it might be good to put a sentence somewhere in the docs to recommend users to set names for multi-index levels before creating data arrays or datasets. What do you think?

File: xarray/core/variable.py (hunk 18 above)
Comment (benbovy) [inline xarray/core/variable.py]: Yes it makes sense and it is more compact. However, how do we handle cases where two or more multi-index dimensions both collapse to `level_0` after indexing? Maybe such cases are rather edge cases so we don't need to care too much about this? Also a dimension / coordinate renamed to `level_0` is not very informative. Maybe we can include the old dimension name when renaming in `_replace_indexes`? Or, if we recommend xarray users to set explicit names for multi-index levels, we can probably leave this as-is...

I let you decide.

**Comment by tippetts** [comment]
Mind if I ask if this will get merged into master? It looks like a lot of work went into the pull request, and the discussion + passed checks lead me to believe it could be close to going in. Is there anything a third party can do to push it across the finish line?

**Comment by shoyer** [comment]
@tippetts Thanks for checking in. I actually hadn't noticed the last round of changes here, but I think this is good to go in now! Big thanks to @benbovy!!!

**Comment by benbovy** [comment]
Thanks again @shoyer for your reviews and comments!

Merging this PR broke the tests added in #881. I fixed that in #903.
